### PR TITLE
fix: check for exception when parsing pub sub response and return it (#2696)

### PR DESCRIPTION
**Pull request by ldetmer**
* fix: check for exception when parsing pub sub response and return it (#2696 )

### CODE REVIEW

File: spring-cloud-gcp-pubsub/src/test/java/com/google/cloud/spring/pubsub/core/subscriber/PubSubSubscriberTemplateTests.java [hunk 1]
```diff
@@ -530,6 +531,21 @@ void testPullAndConvertAsync()
...
+  @Test
+  void testPullAndConvertAsync_publishesConvertException()
+          throws InterruptedException, ExecutionException, TimeoutException {
+    when(this.messageConverter.fromPubSubMessage(this.pubsubMessage, BigInteger.class)).thenThrow(new NullPointerException());
+    CompletableFuture<List<ConvertedAcknowledgeablePubsubMessage<BigInteger>>> asyncResult =
+            this.pubSubSubscriberTemplate.pullAndConvertAsync("sub2", 1, true, BigInteger.class);
+
+    try {
+      asyncResult.get();
+      Assertions.fail();
```
Comment (suztomo) [inline spring-cloud-gcp-pubsub/src/test/java/com/google/cloud/spring/pubsub/core/subscriber/PubSubSubscriberTemplateTests.java:543]: Add a message that appears when this assertion fails.

File: spring-cloud-gcp-pubsub/src/test/java/com/google/cloud/spring/pubsub/core/subscriber/PubSubSubscriberTemplateTests.java [hunk 2]
```diff
@@ -530,6 +531,21 @@ void testPullAndConvertAsync()
...
 
+  @Test
+  void testPullAndConvertAsync_publishesConvertException()
+          throws InterruptedException, ExecutionException, TimeoutException {
+    when(this.messageConverter.fromPubSubMessage(this.pubsubMessage, BigInteger.class)).thenThrow(new NullPointerException());
+    CompletableFuture<List<ConvertedAcknowledgeablePubsubMessage<BigInteger>>> asyncResult =
+            this.pubSubSubscriberTemplate.pullAndConvertAsync("sub2", 1, true, BigInteger.class);
+
+    try {
+      asyncResult.get();
```
Comment (suztomo) [inline spring-cloud-gcp-pubsub/src/test/java/com/google/cloud/spring/pubsub/core/subscriber/PubSubSubscriberTemplateTests.java:542]: Does the test fail if there's no change in spring-cloud-gcp-pubsub/src/main/java/com/google/cloud/spring/pubsub/core/subscriber/PubSubSubscriberTemplate.java?

File: spring-cloud-gcp-pubsub/src/test/java/com/google/cloud/spring/pubsub/core/subscriber/PubSubSubscriberTemplateTests.java (hunk 1 above)
Comment (ldetmer) [reply spring-cloud-gcp-pubsub/src/test/java/com/google/cloud/spring/pubsub/core/subscriber/PubSubSubscriberTemplateTests.java]: done

File: spring-cloud-gcp-pubsub/src/test/java/com/google/cloud/spring/pubsub/core/subscriber/PubSubSubscriberTemplateTests.java [hunk 3]
```diff
@@ -530,6 +531,21 @@ void testPullAndConvertAsync()
...
+      Assertions.fail("NullPointerException should have been thrown and was not!");
```
Comment (burkedavison) [inline spring-cloud-gcp-pubsub/src/test/java/com/google/cloud/spring/pubsub/core/subscriber/PubSubSubscriberTemplateTests.java:543]: No change requested, just a consideration: Another way to write this check is to use the `assertThrows` function which has the benefit of providing an automatic clear test failure synopsis.

```java
ExecutionException e = assertThrows(ExecutionException.class, () -> asyncResult.get());
assertThat(e.getCause()).isInstanceOf(NullPointerException.class);
```

**Comment by burkedavison** [comment]
LGTM once Tomo gives the thumbs up.

File: spring-cloud-gcp-pubsub/src/test/java/com/google/cloud/spring/pubsub/core/subscriber/PubSubSubscriberTemplateTests.java (hunk 2 above)
Comment (ldetmer) [reply spring-cloud-gcp-pubsub/src/test/java/com/google/cloud/spring/pubsub/core/subscriber/PubSubSubscriberTemplateTests.java]: Good catch, updated to add a timeout to fail in the case that the response hangs (no change to PubSubSubscriberTemplate)

File: spring-cloud-gcp-pubsub/src/test/java/com/google/cloud/spring/pubsub/core/subscriber/PubSubSubscriberTemplateTests.java (hunk 3 above)
Comment (ldetmer) [reply spring-cloud-gcp-pubsub/src/test/java/com/google/cloud/spring/pubsub/core/subscriber/PubSubSubscriberTemplateTests.java]: nice that's cleaner! applied suggestion

File: spring-cloud-gcp-pubsub/src/test/java/com/google/cloud/spring/pubsub/core/subscriber/PubSubSubscriberTemplateTests.java [hunk 4]
```diff
@@ -530,6 +533,17 @@ void testPullAndConvertAsync()
     assertThat(result.get(0).getProjectSubscriptionName().getSubscription()).isEqualTo("sub2");
   }
 
+  @Test
+  @Timeout(value = 5)
```
Comment (suztomo) [inline spring-cloud-gcp-pubsub/src/test/java/com/google/cloud/spring/pubsub/core/subscriber/PubSubSubscriberTemplateTests.java:537]: Would you add a unit? Without a unit, I had to look up the API reference (https://junit.org/junit5/docs/5.5.1/api/org/junit/jupiter/api/Timeout.html).

**Review by suztomo — APPROVED** [review]
Would you add a unit and them merge this?

**Review by burkedavison — APPROVED** [review]

**Pull request merged**